--- include/machine/asm.h	2025-06-15 09:19:38 +0200
+++ include/machine/asm.h	2025-06-15 09:29:56 +0200
@@ -58,7 +58,7 @@
 #define PIC_GOTOFF(x)	x
 #endif
 
-#define _C_LABEL(name)	name
+#define _C_LABEL(name)	__CONCAT(_,name)
 #define	_ASM_LABEL(x)	x
 
 #define CVAROFF(x, y)	_C_LABEL(x) + y
@@ -94,7 +94,7 @@
 #endif
 
 #define _ENTRY(x) \
-	.text; _ALIGN_TEXT; .globl x; .type x,@function; x:
+	.text; _ALIGN_TEXT; .global x; .def x; .scl 2; .type 32; .endef; x:
 
 #if defined(PROF) || defined(GPROF)
 # define _PROF_PROLOGUE	\
--- include/machine/cdefs.h	2025-06-15 09:19:38 +0200
+++ include/machine/cdefs.h	2025-06-15 09:29:52 +0200
@@ -4,18 +4,81 @@
  * Written by J.T. Conklin <jtc@wimsey.com> 01/17/95.
  * Public domain.
  */
+/* Adapted for MinC by D. Boland <dboland@xs4all.nl> 2016
+ */
 
 #ifndef	_MACHINE_CDEFS_H_
 #define	_MACHINE_CDEFS_H_
 
-#define __strong_alias(alias,sym)			\
-	__asm__(".global " __STRING(alias) " ; "	\
-	    __STRING(alias) " = " __STRING(sym))
-#define __weak_alias(alias,sym)				\
-	__asm__(".weak " __STRING(alias) " ; "		\
-	    __STRING(alias) " = " __STRING(sym))
-#define __warn_references(sym,msg)			\
-	__asm__(".section .gnu.warning." __STRING(sym)	\
-	    " ; .ascii \"" msg "\" ; .text")
+#ifndef __bounded__
+#define __bounded__(...)	
+#endif
+#ifndef visibility
+#define visibility(a)	
+#endif
+
+#undef __MINGW32__
+#undef __WIN32__	// <GL/gl.h>
+#undef __MSVCRT__
+#undef __WINNT__
+#undef __WINNT
+#undef _WIN32		// cc.exe
+#undef WIN32		// cc.exe
+#undef WINNT
+
+#define __thread	__thread __attribute__((section(".tls")))
+#define __import	__attribute__((dllimport))
+
+#define __MinC__
+#define __OpenBSD__
+#define restrict	__restrict
+#define environ		__environ
+#define alloca		__builtin_alloca
+
+#define sys_siglist	_sys_siglist	/* legacy */
+
+/* pthread_win32 */
+#define PTW32_CDECL
+#define __struct_timespec_defined
+#define __have_typedef_pid_t
+
+#ifdef USEVFORK
+#define fork		vfork
+#endif
+
+#ifdef NOPIE
+#define LIBC_IMPORT		
+#define GETOPT_IMPORT	
+#define LIBSOCK_IMPORT	
+#define DLL_IMPORT	
+#endif
+
+#ifndef LIBC_IMPORT
+#define LIBC_IMPORT	__import
+#endif
+#ifndef GETOPT_IMPORT
+#define GETOPT_IMPORT	__import
+#endif
+#ifndef LIBSOCK_IMPORT
+#define LIBSOCK_IMPORT	__import
+#endif
+#ifndef DLL_IMPORT
+#define DLL_IMPORT	__import
+#endif
+
+/* weak aliases are kind of broken in MinGW32 ld.exe:
+ * http://mingw.5.n7.nabble.com/weak-functions-td6379.html
+ * https://sourceware.org/bugzilla/show_bug.cgi?id=9687
+ * https://sourceware.org/bugzilla/show_bug.cgi?id=2729
+ */
+
+#define __strong_alias(name,symbol)			\
+	extern __typeof(symbol) name __attribute__((alias(__STRING(symbol))))
+//#define __weak_alias(name,symbol)			\
+//	extern __typeof(symbol) name __attribute__((weak,alias(__STRING(symbol))))
+#define __weak_alias(name,symbol)			\
+	__strong_alias(name,symbol)
+#define __warn_references(symbol,msg)			\
+	extern __typeof(symbol) symbol __attribute__((deprecated(msg)))
 
 #endif /* !_MACHINE_CDEFS_H_ */
--- include/machine/exec.h	2025-06-15 09:19:38 +0200
+++ include/machine/exec.h	2025-06-15 09:29:46 +0200
@@ -40,7 +40,6 @@
 #define ELF_TARG_MACH		EM_386 /* XXX - EM_486 is currently unused
                                           by all OSs/compilers/linkers */
 
-#define _NLIST_DO_ELF
 #define _KERN_DO_ELF
 
 #endif  /* _MACHINE_EXEC_H_ */
--- include/machine/_types.h	2025-06-15 09:19:38 +0200
+++ include/machine/_types.h	2025-06-15 09:30:01 +0200
@@ -121,7 +121,7 @@
 typedef long double		__double_t;
 typedef long double		__float_t;
 typedef long			__ptrdiff_t;
-typedef	unsigned long		__size_t;
+typedef	__SIZE_TYPE__		__size_t;
 typedef	long			__ssize_t;
 #if defined(__GNUC__) && __GNUC__ >= 3
 typedef	__builtin_va_list	__va_list;
@@ -131,9 +131,9 @@
 
 /* Wide character support types */
 #ifndef __cplusplus
-typedef	int			__wchar_t;
+typedef	__WCHAR_TYPE__	__wchar_t;		// match WinNT's
 #endif
-typedef int			__wint_t;
+typedef	__WINT_TYPE__		__wint_t;		// match WinNT's
 typedef	int			__rune_t;
 typedef	void *			__wctrans_t;
 typedef	void *			__wctype_t;
